src/nimblepkg/checksums.nim

```diff
@@ -0,0 +1,50 @@
+# Copyright (C) Dominik Picheta. All rights reserved.
+# BSD License. Look at license.txt for more info.
+
+import os, std/sha1, strformat, algorithm
+import common, version, sha1hashes, vcstools, paths
+
+type
+  ChecksumError* = object of NimbleError
+
+proc checksumError*(name: string, version: Version,
+                    vcsRevision, checksum, expectedChecksum: Sha1Hash):
+    ref ChecksumError =
+  result = newNimbleError[ChecksumError](&"""
+Downloaded package checksum does not correspond to that in the lock file:
+  Package:           {name}@v.{version}@r.{vcsRevision}
+  Checksum:          {checksum}
+  Expected checksum: {expectedChecksum}
+""")
+
+proc updateSha1Checksum(checksum: var Sha1State, fileName, filePath: string) =
+  if not filePath.fileExists:
+    # In some cases a file name returned by `git ls-files` or `hg manifest`
+    # could be an empty directory name and if so trying to open it will result
+    # in a crash. This happens for example in the case of a git sub module
+    # directory from which no files are being installed.
+    return
+  checksum.update(fileName)
+  let file = filePath.open(fmRead)
+  defer: close(file)
+  const bufferSize = 8192
+  var buffer = newString(bufferSize)
+  while true:
+    var bytesRead = readChars(file, buffer)
+    if bytesRead == 0: break
+    checksum.update(buffer[0..<bytesRead])
```

Comment:
What is the alternative?

Reply:
I prefer using the asynctools library because it enables us in the future to implement some parallelization of downloads when doing recursive downloads without using a lock file. This is not possible with the `osproc.execProcesses`.

Reply:
Well, I made it with the `toOpenArray` procedure to avoid the copy.